crates/wasm-mutate/src/mutators/peephole/swap_commutative.rs

```diff
@@ -0,0 +1,184 @@
+                local_count += count;
+                (count, map_type(ty).unwrap())
+            })
+            .collect::<Vec<(u32, ValType)>>();
+
+        // add two temporary locals, the last two
+        let (operator, _) = &operators[operator_index];
+        current_locals.push((2, self.get_operator_type(operator)?));
+
+        let mut newf = Function::new(current_locals);
+        let mut idx = 0;
+
+        let mut newoffset = 0;
+        for (_, offset) in operators {
+            newoffset = offset;
+            if idx == operator_index {
+                // Copy previous code to the body
+                let previous = &function_stream[body_range.start..offset];
+                newf.raw(previous.iter().copied());
+                // Inject new code to swap operands
+                newf.instruction(Instruction::LocalSet(local_count + 1));
+                newf.instruction(Instruction::LocalSet(local_count + 2));
+                newf.instruction(Instruction::LocalGet(local_count + 1));
+                newf.instruction(Instruction::LocalGet(local_count + 2));
+                break; // this break allows to copy the remaining buffer of the current reader
+            }
+            idx += 1;
+        }
+
+        // Copy last part of the function body
+        let remaining = &function_stream[newoffset..body_range.end];
+        newf.raw(remaining.iter().copied());
+        Ok(newf)
+    }
+
+    fn can_mutate<'a>(
+        &self,
+        _: &'a WasmMutate,
+        operators: &Vec<TupleType<'a>>,
+        at: usize,
+    ) -> Result<bool> {
+        let (operator, _) = &operators[at];
+        Ok(self.is_commutative(operator))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::{rngs::SmallRng, SeedableRng};
+    use wasm_encoder::{CodeSection, FunctionSection, Module, TypeSection, ValType};
+    use wasmparser::{Chunk, Parser};
+
+    use crate::mutators::peephole::TupleType;
+    use crate::{
+        mutators::{
+            peephole::{CodeMutator, PeepholeMutator},
+            Mutator,
+        },
+        WasmMutate,
+    };
+    use wasm_encoder::{RawSection, SectionId};
+    use wasmparser::{Payload, SectionReader};
+
+    use super::SwapCommutativeOperator;
+
+    #[test]
+    fn test_swap() {
+        let original = r#"
+        (module
+            (func (export "exported_func") (result i32) (local i32 i32)
+                i32.const 42
+                i32.const 42
+                i32.add
+                i32.const 56
+                i32.add
+            )
+        )
+        "#;
+
+        let expected = r#"
+        (module
+            (type (;0;) (func (result i32)))
+            (func (;0;) (type 0) (result i32)
+              (local i32 i32 i32 i32)
+              i32.const 42
+              i32.const 42
+              i32.add
+              i32.const 56
+              local.set 3
+              local.set 4
+              local.get 3
+              local.get 4
+              i32.add))
+        "#;
```

Comment:
Why isn't the function exported anymore after the mutation? Swapping the operands of a commutative operation shouldn't have any effect on whether the operation's function is exported or not.

Reply:
Right...this is on the test macro..I was not saving all the sections. I will fix this :)

Reply:
Done